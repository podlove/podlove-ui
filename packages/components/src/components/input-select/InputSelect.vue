<script lang="ts" setup>
import { path } from 'ramda';

defineProps({
  disabled: {
    type: Boolean,
    default: false
  },
  options: {
    type: Array,
    default: () => []
  },
  value: {
    type: String,
    default: null
  }
});

const emit = defineEmits(['change']);

const changeEvent = (event: Event) => {
  emit('change', path(['target', 'value'], event));
};
</script>

<template>
  <select
    v-if="options"
    class="podlove-component-select block w-full border rounded-none p-1 h-8"
    :disabled="disabled"
    :value="value"
    @change="changeEvent"
  >
    <option v-for="(option, index) in options" :key="index" :selected="option === value">
      {{ option }}
    </option>
  </select>
  <select v-else class="podlove-component-select" :disabled="disabled" @change="changeEvent">
    <slot />
  </select>
</template>

<style lang="postcss" scoped>
.podlove-component-select {
  color: var(--podlove-component--input-select--color, var(--podlove-components-background));
  background: var(--podlove-component--input-select--background, var(--podlove-components-text));
  border-color: var(--podlove-component--input-select--border, var(--podlove-components-background));
}
</style>
