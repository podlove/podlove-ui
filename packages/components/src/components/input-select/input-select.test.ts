import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import InputSelect from './InputSelect.vue';

describe('InputSelect', () => {
  describe('component', () => {
    test('should be a Vue instance', () => {
      const wrapper = mount(InputSelect);
      expect(wrapper.vm).toBeTruthy();
    });
  });

  describe('props', () => {
    test('should render with default props', () => {
      const wrapper = mount(InputSelect);
      expect(wrapper.element).toMatchSnapshot();
    });

    const props = [
      {
        prop: 'disabled',
        value: true
      },
      {
        prop: 'color',
        value: '#fff'
      },
      {
        prop: 'background',
        value: '#fff'
      },
      {
        prop: 'borderColor',
        value: '#fff'
      },
      {
        prop: 'value',
        value: 'foo'
      }
    ];

    props.forEach(({ prop, value }) => {
      test(`should render type '${prop}'`, async () => {
        const wrapper = await mount(InputSelect, {
          props: {
            [prop]: value,
            options: ['foo', 'bar', 'baz']
          }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('events', () => {
    test('should dispatch change event on change', async () => {
      const wrapper = await mount(InputSelect, {
        props: {
          options: ['foo', 'bar', 'baz'],
          value: 'bar'
        }
      });

      await wrapper.trigger('change');
      expect(wrapper.emitted().change).toEqual([['bar']]);
    });
  });
});
