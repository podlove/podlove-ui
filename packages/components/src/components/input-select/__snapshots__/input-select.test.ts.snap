// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputSelect > props > should render type 'background' 1`] = `
<select
  background="#fff"
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  data-v-729933d8=""
>
  
  <option
    data-v-729933d8=""
  >
    foo
  </option>
  <option
    data-v-729933d8=""
  >
    bar
  </option>
  <option
    data-v-729933d8=""
  >
    baz
  </option>
  
</select>
`;

exports[`InputSelect > props > should render type 'borderColor' 1`] = `
<select
  bordercolor="#fff"
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  data-v-729933d8=""
>
  
  <option
    data-v-729933d8=""
  >
    foo
  </option>
  <option
    data-v-729933d8=""
  >
    bar
  </option>
  <option
    data-v-729933d8=""
  >
    baz
  </option>
  
</select>
`;

exports[`InputSelect > props > should render type 'color' 1`] = `
<select
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  color="#fff"
  data-v-729933d8=""
>
  
  <option
    data-v-729933d8=""
  >
    foo
  </option>
  <option
    data-v-729933d8=""
  >
    bar
  </option>
  <option
    data-v-729933d8=""
  >
    baz
  </option>
  
</select>
`;

exports[`InputSelect > props > should render type 'disabled' 1`] = `
<select
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  data-v-729933d8=""
  disabled=""
>
  
  <option
    data-v-729933d8=""
  >
    foo
  </option>
  <option
    data-v-729933d8=""
  >
    bar
  </option>
  <option
    data-v-729933d8=""
  >
    baz
  </option>
  
</select>
`;

exports[`InputSelect > props > should render type 'value' 1`] = `
<select
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  data-v-729933d8=""
>
  
  <option
    data-v-729933d8=""
  >
    foo
  </option>
  <option
    data-v-729933d8=""
  >
    bar
  </option>
  <option
    data-v-729933d8=""
  >
    baz
  </option>
  
</select>
`;

exports[`InputSelect > props > should render with default props 1`] = `
<select
  class="podlove-component-select block w-full border rounded-none p-1 h-8"
  data-v-729933d8=""
>
  
  
</select>
`;
